synapse/handlers/e2e_keys.py

```diff
@@ -155,16 +160,26 @@ async def query_devices(self, query_body, timeout, from_user_id):
                             unsigned["device_display_name"] = device_display_name
                         user_devices[device_id] = result
 
+            # check for missing cross-signing keys.
+            for user_id in remote_queries.keys():
+                cached_cross_master = user_id in cross_signing_keys["master_keys"]
+                cached_cross_selfsigning = (
+                    user_id in cross_signing_keys["self_signing_keys"]
+                )
+
+                # check if only one of the cross-signing master and
+                # self-signing key are cached.
+                # for each user we want the master _and_ the self-signing key,
+                # so we fetch those keys from federation
+                if cached_cross_master ^ cached_cross_selfsigning:
+                    user_ids_not_in_cache.add(user_id)
```

Comment:
it would be good to include a comment here about why we do not do it if we have *both* keys. (As I understand it: to avoid making a federation request for every device query for a user who has not enabled cross-signing)

Reply:
indeed. i'm not sure how the signaling works when a user gains new (or initial) cross-signing keys, but that is hopefully handled correctly some way already.